internal/storage/download_test.go

```diff
@@ -181,119 +222,67 @@ func TestDownloadToFileError(t *testing.T) {
-func host(addr string, name string, size int64, simple bool, t *testing.T) *web {
-	// Create new HTTP server.
-	w := &web{name: name, size: size, srv: &http.Server{Addr: addr}, t: t}
-
-	// Add HTTP alias only once.
-	if simple {
-		testOnceSimple.Do(func() { http.HandleFunc(fmt.Sprintf("/%s", name), w.handlerSimple) })
-	} else {
-		testOnceRange.Do(func() { http.HandleFunc(fmt.Sprintf("/%s", name), w.handlerRange) })
+// TestDownloadToFileSecureError tests HTTPS file download function for bad/expired TLS certificates.
+func TestDownloadToFileSecureError(t *testing.T) {
+	// Prepare
+	dir := "_tmp-download"
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		t.Fatalf("failed create temporary directory: %v", err)
 	}
 
-	// Start HTTP server in separate goroute.
-	go func() {
-		if err := w.srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
-			t.Errorf("failed to start web server: %v", err)
-		}
-	}()
-
-	time.Sleep(1 * time.Second)
-
-	return w
-}
+	// Remove temporary directory at the end
+	defer os.RemoveAll(dir)
 
-// close the http server.
-func (w *web) close() {
-	if err := w.srv.Shutdown(context.Background()); err != nil {
-		w.t.Errorf("failed shutdown web server: %v", err)
+	art := &Artifact{
+		FileName: "test.txt", Size: 65536,
+		HashType:  "MD5",
+		HashValue: "ab2ce340d36bbaafe17965a3a2c6ed5b",
 	}
-}
 
-// handlerRange handles incoming HTTP requests with range header support.
-func (w *web) handlerRange(writer http.ResponseWriter, request *http.Request) {
-	// Set default headers for txt file.
-	writer.Header().Set("Content-Type", "text/plain")
-	writer.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, w.name))
-	writer.Header().Set("Accept-Ranges", "bytes")
-
-	// Parse range header -> 'Range: bytes=<begin>-<end>'
-	r := request.Header.Get("range")
-	if r == "" { // No range is defined
-		writer.Header().Set("Content-Length", strconv.Itoa(int(w.size)))
-		write(writer, w.size)
-		return
-	}
+	// Start Web servers secure
+	srvSecureInv := Host(":43234", art.FileName, int64(art.Size), true, true, expiredCert, expiredKey, t)
+	defer srvSecureInv.Close()
+	srvSecureExt := Host(":43235", art.FileName, int64(art.Size), true, true, extCert, extKey, t)
+	defer srvSecureExt.Close()
+	srvSecureVal := Host(":43236", art.FileName, int64(art.Size), true, true, validCert, validKey, t)
+	defer srvSecureVal.Close()
+	name := filepath.Join(dir, art.FileName)
 
-	// Client requests a part of the file
-	var err error
-	begin, end := int64(0), int64(w.size-1)
-	r = r[6:] // Strip the "bytes=", left over is now "begin-end"
-	pos := strings.Index(r, "-")
-	if pos < 0 {
-		http.Error(writer, "invalid values for header 'Range'", http.StatusBadRequest)
-		return
+	// 1. Try download with expired certificate
+	art.Link = "https://localhost:43234/test.txt"
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
+		t.Fatalf("validated with expired certificate: %v", err)
```

Comment:
The message can be done more specific for this case where the client is not using the expired certificate.